tt_metal/common/blockfloat_common.hpp

```diff
@@ -0,0 +1,258 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#include <algorithm>
+#include <iostream>
+#include <random>
+#include <vector>
+#include <immintrin.h>
+
+#include "common/assert.hpp"
+#include "common/logger.hpp"
+#include "common/tt_backend_api_types.hpp"
+#include "tt_metal/third_party/tracy/public/tracy/Tracy.hpp"
+
+using namespace std;
+
+inline uint8_t get_max_exp(const vector<uint32_t> &vec, bool is_exp_a) {
+    TT_ASSERT(vec.size() == 16);
+    uint32_t max = 0;
+
+    for (int i = 0; i < 16; ++i) {
+        // mask & shift out exp
+        uint32_t exp = (vec[i] & 0x7f800000) >> 23;
+
+        if (is_exp_a) {
+            int32_t se = static_cast<int32_t>(exp);
+            // need to rebias from 127 to 15
+            se = se - 127 + 15;
+
+            if (se > 31) {
+                se = 31;
+            }
+            else if (se < 0) {
+                se = 0;
+            }
+
+            exp = static_cast<uint32_t>(se);
+        }
+
+        if (exp > max) {
+            max = exp;
+        }
+    }
+    return max;
+}
+
+inline uint32_t get_exp_dword(const vector<uint8_t> &vec) {
+    TT_ASSERT(vec.size() == 4);
+
+    uint32_t tmp = 0;
+    for(int i = 0; i < 4; ++i) {
+        tmp = tmp | ((vec[i] & 0xff) << (i*8));
+    }
+    return tmp;
+}
+
+inline uint32_t get_byte(uint32_t word, uint32_t index) {
+    TT_ASSERT(index < 4);
+    uint32_t mask = 0xff << (8 * index);
+    uint32_t masked = word & mask;
+    masked = masked >> (8 * index);
+    return masked;
+}
+
+template <tt::DataFormat BfpFormat, bool truncate_bfp_mantissa=false>
+inline uint8_t convert_u32_to_bfp(uint32_t input, uint32_t shared_exp, bool is_exp_a) {
```

Comment:
@ttmtrajkovic nice work commonizing all the block float logic here.. do you think we could convert the existing `convert_u32_to_bfp8` to use your function here as well?

Reply:
do you think we could create a template parameter to configure the number of round bits similar to what you did with `bool truncate_bfp_mantissa=false`? Borrowed idea from BUDA is that for bfp8_b, you might want to round at bit 5 when using LoFi because it can give better accuracy?

Reply:
```cpp
            if(round_bfp_to_5_bits) {
                std::uint32_t round_bit_pos = std::min(5 + exp_diff, 7); // Optimal matinssa rounding bit, based on how far exp for this mantissa is from max exp (need more precision for smaller numbers)
                m += 1 << (23 - round_bit_pos); // 5 bit round: 1 << 18, 6 bit round: 1 << 17, 7 bit round (default) : 1 << 16
                m = m >> (24 - round_bit_pos); // 5 bit round: >> 19, 6 bit round: >> 18, 7 bit round: >> 17
                m = m << (7 - round_bit_pos); // Left shift to be consistent with exp
                if(m > 127) m = 127;
            }
```

Reply:
re rounding, yes, that could be done although rounding a bfp mantissa isn't exactly the same as chopping off bits for lofi.
When converting to bfp, since 16 numbers share the same exponent, mantissas are shifted so it is possible that rounding to 5bits only we actually lose some of the bits that actually make the difference during lofi math.
I can explain this scenario further if you want in a chat, it might be that I'm missing something

With float16 and lofi, it does make sense to round to 5bits rather then 7 as there's no mantissa shifting.

Reply:
With bfp_common file, we should be able to easilly use this for bfp8, and eventually bfp2 if we need it.
I will submit this as is

Reply:
@cfjchu, do you think its ok to submit like this? if yes, could you approve?

Reply:
I'll follow-up with you offline. Everything else looks great otherwise!